### PR TITLE
Fix entropy calculation in SFT

### PR DESCRIPTION
**Pull request by qgallouedec**
SFT entropy was averaged over the wrong tokens

In SFT, the model is trained to predict the **next token** at positions where `label != -100`. The `entropy` metric is supposed to measure the model's uncertainty on **those same positions**. Before this PR it didn't.

Throughout this note I'll use one running example: a sequence of 6 tokens, 3 prompt tokens + 3 completion tokens, under `completion_only_loss=True` (prompt is masked out of the loss).

```text
position     :   0     1     2     3     4     5
token        :   P₁    P₂    P₃    A     B     C
label        : -100  -100  -100    A     B     C     ← -100 means "don't train here"
entropy[pos] :  2.0   2.2   1.8   1.5   1.0   1.3    ← model's uncertainty at that position
```

## Bug 1 — Misaligned by one position

Next-token training: `logits[i]` predicts `token[i+1]`. The loss pairs them like this:

```text
 logits[0] → label[1]     (prompt, masked: -100)
 logits[1] → label[2]     (prompt, masked: -100)
 logits[2] → label[3] = A  ← loss trains on this pair
 logits[3] → label[4] = B  ← loss trains on this pair
 logits[4] → label[5] = C  ← loss trains on this pair
 logits[5] → (no next)      dropped
```

The loss looks at `entropy[2], entropy[3], entropy[4]`. The **old entropy metric didn't shift**, so it looked at `entropy[0..5]` — a different set of positions.

## Bug 2 — Averaged over the wrong tokens

Even with the shift fixed, the old code weighted by `attention_mask` (which is `1` for every real, non-padded token). Under `completion_only_loss`, prompt tokens are real but `label == -100` — they're excluded from the loss but were included in entropy:

```text
positions used by loss:         2, 3, 4              → entropy over {1.8, 1.5, 1.0}  ← correct
positions used by old metric:   0, 1, 2, 3, 4, 5     → entropy over all six
```

Correct metric: `(1.8 + 1.5 + 1.0) / 3 = 1.43`
Old metric (full attention_mask, no shift): `(2.0 + 2.2 + 1.8 + 1.5 + 1.0 + 1.3) / 6 = 1.63`

The old number is inflated by the prompt tokens' entropy, which tells you nothing about how well training is going.

## Bug 3 — Wrong aggregation across GPUs

Now imagine the sequence is split across 2 GPUs by sequence parallelism:

```text
rank 0 (positions 0-2)          rank 1 (positions 3-5)
 P₁    P₂    P₃                  A     B     C
-100  -100  -100                 A     B     C
 2.0   2.2   1.8                 1.5   1.0   1.3

valid (label≠-100): 0 tokens     valid: 3 tokens
local entropy      : — (nan/0)    local entropy: (1.5+1.0+1.3)/3 = 1.27
```

**Old aggregation**: take unweighted mean of per-rank entropies → `(0 + 1.27) / 2 = 0.63` — absurd, and drops as soon as you add more ranks of prompt.

**Correct aggregation**: sum counts across ranks, then divide.

```text
 total entropy_sum across ranks : 0 + (1.5+1.0+1.3) = 3.8
 total valid tokens across ranks: 0 + 3             = 3
 entropy                        : 3.8 / 3           = 1.27   ← matches non-SP result
```

## The fix

Entropy is now computed identically to token accuracy (which already did all three things right):

```python
shift_logits  = outputs.logits[..., :-1, :]          # or inputs["shift_labels"] under CP/SP
shift_labels  = labels[..., 1:]
mask          = shift_labels != -100

per_token_entropy = entropy_from_logits(shift_logits)
entropy_sum       = (per_token_entropy * mask).sum()   # raw count
total_tokens      = mask.sum()                         # raw count

# Sum counts across ranks, then divide
entropy_sum   = gather(entropy_sum).sum()
total_tokens  = gather(total_tokens).sum()
entropy       = entropy_sum / total_tokens
```

---

> [!NOTE]
> **Low Risk**
> Low risk because this only changes how the `entropy` metric is computed/aggregated (no loss/backprop changes), but it will alter logged values and could affect monitoring/regression thresholds in distributed setups.
> 
> **Overview**
> Fixes the `entropy` metric computation in `SFTTrainer` to match next-token training positions by *shifting logits/labels consistently*, masking by `label != -100` (instead of `attention_mask`), and correctly aggregating across distributed ranks via *sum/count* weighting.
> 
> Applies the same *sum/count across ranks* aggregation pattern to `DPOTrainer` entropy logging to avoid incorrect unweighted per-rank averaging under sequence parallelism.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 654bb18e62f643e36aeacd97ff90f6a576279514. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by qgallouedec** [comment]
@codex review

**Review by albertvillanova — APPROVED** [review]
Thanks a lot for the fix!
